Add finishAndThrow function to ProcessGroup::Work, and use with Gloo

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40405 Add finishAndThrow function to ProcessGroup::Work, and use with Gloo**
* #40404 [NCCL Docs] Adding Comments for Work-level Finish in ProcessGroup

This adds a finishAndThrow function that completes the work object,
sets an exception if one is provided by the user, and throws an exception (if
it is already set or passed by the caller). This is now done by grabbing the
lock just once and simplifies the wait functions in ProcessGroupGloo.

Differential Revision: [D22174890](https://our.internmc.facebook.com/intern/diff/D22174890/)

osalpekar committed:
Add finishAndThrow function to ProcessGroup::Work, and use with Gloo

This adds a finishAndThrow function that completes the work object,
sets an exception if one is provided by the user, and throws an exception (if
it is already set or passed by the caller). This is now done by grabbing the
lock just once and simplifies the wait functions in ProcessGroupGloo.

Differential Revision: [D22174890](https://our.internmc.facebook.com/intern/diff/D22174890/)

[ghstack-poisoned]

osalpekar committed:
Update on "Add finishAndThrow function to ProcessGroup::Work, and use with Gloo"


This adds a finishAndThrow function that completes the work object,
sets an exception if one is provided by the user, and throws an exception (if
it is already set or passed by the caller). This is now done by grabbing the
lock just once and simplifies the wait functions in ProcessGroupGloo.

Differential Revision: [D22174890](https://our.internmc.facebook.com/intern/diff/D22174890/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -341,14 +341,9 @@ bool ProcessGroupGloo::SendWork::wait() {
-  }
+
+  // Completes the Work object and throws the exception.
+  finishAndThrow(exception);
```
curious, why this cannot be moved to the `catch` block, saving a `std::exception_ptr` var?

osalpekar replied:
It probably can be moved inside - just that we still need to mark the work object as completed even if there's no exception. Since this logic sequence (mark complete, set exception, then throw if needed) is pretty common across the ProcessGroup backends, it would be nice to have a helper function that does all this and only grabs the lock once.

jiayisuse commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -56,6 +56,15 @@ void ProcessGroup::Work::finish(std::exception_ptr exception) {
+void ProcessGroup::Work::finishAndThrow(std::exception_ptr exception) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  completed_ = true;
+  exception_ = exception;
```
Could nullptr `exception `overwrite` exception_`, hence we miss the previous exception?

osalpekar replied:
Yes, this could happen, but this is possible in the gloo case - if no exception is caught by `waitSend/waitRecv`, then `exception_` is set to `nullptr` and nothing is thrown, even if an exception was previously set.

We can remove the `nullptr` from being the default value for the `exception_ptr` argument, so the user has to explicitly pass in a `nullptr` if they want to overwrite the last exception.

osalpekar committed:
Update on "Add finishAndThrow function to ProcessGroup::Work, and use with Gloo"


This adds a finishAndThrow function that completes the work object,
sets an exception if one is provided by the user, and throws an exception (if
it is already set or passed by the caller). This is now done by grabbing the
lock just once and simplifies the wait functions in ProcessGroupGloo.

Differential Revision: [D22174890](https://our.internmc.facebook.com/intern/diff/D22174890/)

[ghstack-poisoned]
